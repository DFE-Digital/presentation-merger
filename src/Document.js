import { set } from "shvl";

import Presentation from "../src/Presentation";

export default class Document {
  constructor() {
    this.slides = [];
    this.doc = {
      "office:document-content": {
        "xmlns:dom": "http://www.w3.org/2001/xml-events",
        "xmlns:draw": "urn:oasis:names:tc:opendocument:xmlns:drawing:1.0",
        "xmlns:fo":
          "urn:oasis:names:tc:opendocument:xmlns:xsl-fo-compatible:1.0",
        "xmlns:presentation":
          "urn:oasis:names:tc:opendocument:xmlns:presentation:1.0",
        "xmlns:script": "urn:oasis:names:tc:opendocument:xmlns:script:1.0",
        "xmlns:smil":
          "urn:oasis:names:tc:opendocument:xmlns:smil-compatible:1.0",
        "xmlns:style": "urn:oasis:names:tc:opendocument:xmlns:style:1.0",
        "xmlns:svg": "urn:oasis:names:tc:opendocument:xmlns:svg-compatible:1.0",
        "xmlns:table": "urn:oasis:names:tc:opendocument:xmlns:table:1.0",
        "xmlns:text": "urn:oasis:names:tc:opendocument:xmlns:text:1.0",
        "xmlns:xlink": "http://www.w3.org/1999/xlink",
        "xmlns:office": "urn:oasis:names:tc:opendocument:xmlns:office:1.0"
      }
    };
  }

  merge(pres) {
    if (!(pres instanceof Presentation)) {
      throw new TypeError(
        `expected Presentation class got ${pres.constructor.name}`
      );
    }
    this.slides = this.slides.concat(pres.slides);
    let slides = this.slides.map(slidesDocument => slidesDocument.content);
    let styles = this.slides.map(slidesDocument => slidesDocument.styles);

    set(
      this.doc,
      [
        "office:document-content",
        "office:body",
        "office:presentation",
        "draw:page"
      ],
      slides
    );
    set(
      this.doc,
      ["office:document-content", "office:automatic-styles", "style:style"],
      styles
    );
  }
}
