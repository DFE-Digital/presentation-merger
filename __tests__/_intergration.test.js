import path from 'path';
import fs from 'fs';
import decompress from 'decompress';
import Document from '../src/Document';

async function mergeFiles(files, outPath) {
  const stream = fs.createWriteStream(outPath, { flags: 'w' });
  const doc = new Document();
  for (let file of files) {
    await doc.mergeFile(file);
  }
  doc.pipe(stream);
  return new Promise((done, fail) => {
    doc.on('end', done);
    doc.on('error', fail);
  });
}

describe('Intergration', () => {
  it('Merges two documents together into one document', async () => {
    const filePath = path.join(__dirname, '__fixtures__/output.odp');

    await mergeFiles(
      [
        path.join(__dirname, '__fixtures__/file1.odp'),
        path.join(__dirname, '__fixtures__/file2.odp')
      ],
      filePath
    );

    const files = await decompress(filePath);

    expect(files).toContainEqual(
      expect.objectContaining({
        path: 'content.xml',
        type: 'file'
      })
    );

    expect(files).toContainEqual(
      expect.objectContaining({
        path: 'META-INF/manifest.xml',
        type: 'file'
      })
    );

    expect(files).toContainEqual(
      expect.objectContaining({
        path: 'mimetype',
        type: 'file'
      })
    );

    expect(
      files.find(f => f.path === 'META-INF/manifest.xml').data.toString()
    ).toMatchSnapshot();
  });
});
