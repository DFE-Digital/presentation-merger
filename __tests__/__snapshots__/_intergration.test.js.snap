// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Intergration Merges two documents together into one document 1`] = `
"<manifest:manifest
  xmlns:manifest=\\"urn:oasis:names:tc:opendocument:xmlns:manifest:1.0\\"
  manifest:version=\\"1.2\\"
>
  <manifest:file-entry
    manifest:full-path=\\"/\\"
    manifest:media-type=\\"application/vnd.oasis.opendocument.presentation\\"
    manifest:version=\\"1.2\\"
  />
  <manifest:file-entry
    manifest:full-path=\\"content.xml\\"
    manifest:media-type=\\"text/xml\\"
  />
  <manifest:file-entry
    manifest:full-path=\\"styles.xml\\"
    manifest:media-type=\\"text/xml\\"
  />
  <manifest:file-entry
    manifest:full-path=\\"Presentation0-Thumbnails/thumbnail.png\\"
    manifest:media-type=\\"image/png\\"
  />
  <manifest:file-entry
    manifest:full-path=\\"Presentation0-Pictures/100000000000032000000258E080B12F.jpg\\"
    manifest:media-type=\\"image/jpeg\\"
  />
  <manifest:file-entry
    manifest:full-path=\\"Presentation0-Pictures/100002010000028200000040779DD47E.png\\"
    manifest:media-type=\\"image/png\\"
  />
  <manifest:file-entry
    manifest:full-path=\\"Presentation0-Pictures/1000000000000040000000400142E835.png\\"
    manifest:media-type=\\"image/png\\"
  />
  <manifest:file-entry
    manifest:full-path=\\"Presentation1-Thumbnails/thumbnail.png\\"
    manifest:media-type=\\"image/png\\"
  />
  <manifest:file-entry
    manifest:full-path=\\"Presentation1-Pictures/2000023600002305000012CD3628FC8C.wmf\\"
    manifest:media-type=\\"\\"
  />
  <manifest:file-entry
    manifest:full-path=\\"Presentation1-Pictures/1000000000000040000000400142E835.png\\"
    manifest:media-type=\\"image/png\\"
  />
</manifest:manifest>
"
`;
